### PR TITLE
[InMemoryDataset redesign] read_many_slices_param

### PR DESCRIPTION
**Pull request by crusaderky**
- Part of #386 
- Blocked by and incorporates #388 
- Blocked by and incorporates #373

Add new function `read_many_slices_params_nd()`.

This internally uses a helper function `IndexChunkMapper.read_many_slices_param()`, which duplicates functionality from `IndexChunkMapper.chunk_submap()`and will eventually replace it once `as_subchunk_map()` is no longer used.

### CODE REVIEW

File: versioned_hdf5/subchunk_map.py [hunk 1]
```diff
@@ -377,6 +566,91 @@ def chunk_submap(
...
+    def read_many_slices_params(
+        self,
+    ) -> tuple[NDArray[np_hsize_t], NDArray[np_hsize_t] | None]:
+        n_sel_chunks = len(self.chunk_indices)
+        max_rows = len(self.idx)
+
+        slices = np.empty((max_rows, 5), dtype=np_hsize_t)
+        slices_v: hsize_t[:, :] = slices
+
+        assert max_rows >= n_sel_chunks
```
Comment (peytondmurray) [inline versioned_hdf5/subchunk_map.py:579]: Some chunks require more than one row :+1:

File: versioned_hdf5/subchunk_map.py [hunk 2]
```diff
@@ -538,6 +903,254 @@ def _maybe_array_idx_to_slice(idx: Any):  # -> NDArray[np_hsize_t] | slice:
...
+    ndim = chunk_idxidx.shape[1]
+
+    # Store the slices in a C array (in the stack) of views (in the heap).
+    # In order to keep the reference counter of the views, which are Python objects,
+    # above zero for the duration of the function, we also need to reference them from a
+    # Python object.
+    slices_refs = []
+    # Can't define module-level constants in pure-python mode
+    # https://github.com/numpy/numpy/blob/b0a52b76c2bebe5c237722589d13bf02affa9c43/numpy/core/include/numpy/ndarraytypes.h#L30
+    # NPY_MAXDIMS = 32
```
Comment (peytondmurray) [inline versioned_hdf5/subchunk_map.py:999]: ?!

File: versioned_hdf5/tests/test_subchunk_map.py [hunk 3]
```diff
@@ -187,6 +197,229 @@ def test_chunks_indexer(args):
...
+            slab_offsets_idx += [i] * count
+        slab_offsets = slab_offsets[slab_offsets_idx]
+    assert len(slab_offsets) == len(slices)
+
+    actual_view = actual[None] if actual.shape == () else actual
+
+    # print("=" * 80)
+    # print(f"{idx=} {chunks=} {shape=} {slab_offsets=}")
+    # print(slices)
+
```
Comment (peytondmurray) [inline versioned_hdf5/tests/test_subchunk_map.py:266]:
```suggestion
```

File: versioned_hdf5/tests/test_subchunk_map.py [hunk 4]
```diff
@@ -187,6 +197,229 @@ def test_chunks_indexer(args):
...
+        dst_stride=getitem_slices_nd[:, 4, :],
+    )
+
+    # print("=" * 80)
+    # print(f"{idx=} {chunks=} {shape=} {slab_offsets=}")
+    # print("slab=\n", slab)
+    # print(f"slices_nd=\n{np.asarray(getitem_slices_nd)}")
+    # print("expect=\n", expect)
+    # print("dst=\n", getitem_dst)
+
```
Comment (peytondmurray) [inline versioned_hdf5/tests/test_subchunk_map.py:348]:
```suggestion
```

**Review by peytondmurray — APPROVED** [review]
A few minor changes suggested. The tests are really helpful here, thank you for the extra care you've taken with these. It would be much harder to have confidence here without them!

File: versioned_hdf5/subchunk_map.py (hunk 2 above)
Comment (crusaderky) [reply versioned_hdf5/subchunk_map.py:1001]: defining an array in the stack below (`slices: hsize_t[:, :][32]`) requires knowing the exact size of the array at compile time. in C/C++ this is done through `#define NPY_MAXDIMS 32`. The Cython equivalent is deprecated: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#deprecated-def-if

None of the suggested alternatives work in pure-python mode.

Comment (crusaderky) [reply versioned_hdf5/subchunk_map.py:1001]: https://github.com/cython/cython/issues/6505

**Pull request merged**